django-stubs/contrib/admin/decorators.pyi

```diff
@@ -1,27 +1,50 @@
 from collections.abc import Callable, Sequence
-from typing import Any, TypeVar
+from typing import Any, TypeVar, overload
 
 from django.contrib.admin import ModelAdmin
 from django.contrib.admin.sites import AdminSite
 from django.db.models import Combinable, QuerySet
 from django.db.models.base import Model
 from django.db.models.expressions import BaseExpression
 from django.http import HttpRequest
+from django.utils.functional import _StrPromise
 
-_ModelT = TypeVar("_ModelT", bound=Model)
+_T = TypeVar("_T")
+_Model = TypeVar("_Model", bound=Model)
+_ModelAdmin = TypeVar("_ModelAdmin", bound=ModelAdmin)
+_Request = TypeVar("_Request", bound=HttpRequest)
```

Comment:
The only remaining dilemma I have is whether the request argument should be a TypeVar, or change it to just `HttpRequest` without TypeVar.

It seems most django-stubs code already assumes that requests are always `HttpRequest` and that tends to be sufficient.

Technically most of the time they're `WSGIRequest`, but that seems to be an implementation detail. There is no explicit documentation about `WSGIRequest` in Django docs (barring testing topics and release notes). And `ASGIRequest` is now now rearing its head.

Unless someone else chimes in, I'll trust your judgment here.